api4/plugin.go

```diff
@@ -338,39 +365,68 @@ func parseMarketplacePluginFilter(u *url.URL) (*model.MarketplacePluginFilter, e
 	}
 
 	u, err := url.ParseRequestURI(downloadUrl)
 	if err != nil {
-		c.Err = model.NewAppError("installFromUrl", "api.plugin.install.invalid_url.app_error", nil, "", http.StatusBadRequest)
-		return
+		return nil, model.NewAppError("downloadFromUrl", "api.plugin.install.invalid_url.app_error", nil, "", http.StatusBadRequest)
 	}
 	if !*c.App.Config().PluginSettings.AllowInsecureDownloadUrl && u.Scheme != "https" {
-		c.Err = model.NewAppError("installFromUrl", "api.plugin.install.insecure_url.app_error", nil, "", http.StatusBadRequest)
-		return
+		return nil, model.NewAppError("downloadFromUrl", "api.plugin.install.insecure_url.app_error", nil, "", http.StatusBadRequest)
 	}
 
 	client := c.App.HTTPService.MakeClient(true)
 	client.Timeout = INSTALL_PLUGIN_FROM_URL_HTTP_REQUEST_TIMEOUT
 
 	resp, err := client.Get(downloadUrl)
 	if err != nil {
-		c.Err = model.NewAppError("installFromUrl", "api.plugin.install.download_failed.app_error", nil, err.Error(), http.StatusBadRequest)
-		return
+		return nil, model.NewAppError("downloadFromUrl", "api.plugin.install.download_failed.app_error", nil, err.Error(), http.StatusBadRequest)
 	}
 	defer resp.Body.Close()
 
 	fileBytes, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		c.Err = model.NewAppError("installFromUrl", "api.plugin.install.reading_stream_failed.app_error", nil, err.Error(), http.StatusBadRequest)
-		return
+		return nil, model.NewAppError("downloadFromUrl", "api.plugin.install.reading_stream_failed.app_error", nil, err.Error(), http.StatusBadRequest)
 	}
 
-	manifest, appErr := c.App.InstallPlugin(bytes.NewReader(fileBytes), force)
+	return fileBytes, nil
+}
+
+func verifyPlugin(c *Context, plugin *model.BaseMarketplacePlugin, pluginFileBytes []byte) *model.AppError {
+	for _, signature := range plugin.Signatures {
+		signatureBytes, err := base64.StdEncoding.DecodeString(signature.Signature)
+		if err != nil {
+			return model.NewAppError("verifyPlugin", "api.plugin.signature_decode.app_error", nil, err.Error(), http.StatusInternalServerError)
+		}
+		appErr := c.App.VerifyPlugin(bytes.NewReader(pluginFileBytes), bytes.NewReader(signatureBytes))
+		if appErr == nil {
+			return nil
+		}
+		mlog.Debug("Plugin signature not verified", mlog.String("public key hash", signature.PublicKeyHash), mlog.Err(appErr))
+	}
+	return model.NewAppError("verifyPlugin", "api.plugin.install.verify_plugin.app_error", nil, "", http.StatusInternalServerError)
+}
+
+func saveSignatures(c *Context, plugin *model.BaseMarketplacePlugin) *model.AppError {
+	for count, signature := range plugin.Signatures {
+		signatureBytes, err := base64.StdEncoding.DecodeString(signature.Signature)
+		if err != nil {
+			return model.NewAppError("saveSignatures", "api.plugin.signature_decode.app_error", nil, err.Error(), http.StatusInternalServerError)
+		}
+
+		filePath := filepath.Join(*c.App.Config().PluginSettings.Directory, fmt.Sprintf("%s.%d.sig", plugin.Manifest.Id, count+1))
```

Comment:
Are we storing and verifying plugin signature on sync/startup as part of this pr?

If so should we also verify plugin signature when it's being installed through a cluster notification [here](https://github.com/mattermost/mattermost-server/blob/72965d864a99628b7bd4dfb632a709512e84bef5/app/plugin_install.go#L60)?